### PR TITLE
Add desktop specific RMF attributes

### PR DESCRIPTION
**Pull request by ayoy**
Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).

**Required**:

Task/Issue URL: https://app.asana.com/0/72649045549333/1207774753650441/f
iOS PR: https://github.com/duckduckgo/iOS/pull/3062
macOS PR: https://github.com/duckduckgo/macos-browser/pull/2954
What kind of version bump will this require?: Major/Minor/Patch

**Optional**:

CC: @amddg44 @samsymons 

**Description**:
Add support for customHomePage, pinnedTab, installedMacAppStore, duckPlayerOnboarded
and duckPlayerEnabled attributes on macOS. Attribute matchers have been refactored by introducing
common protocols for matching single value, numeric ranges and string ranges, allowing to remove
a lot of duplicated code as a result.

**Steps to test this PR**:
Verify that unit tests pass.

**OS Testing**:

* [ ] iOS 14
* [ ] iOS 15
* [ ] iOS 16
* [ ] macOS 10.15
* [ ] macOS 11
* [ ] macOS 12

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: Sources/RemoteMessaging/Matchers/AppAttributeMatcher.swift [hunk 1]
```diff
@@ -20,7 +20,59 @@ import Foundation
 import Common
 import BrowserServicesKit
 
-public struct AppAttributeMatcher: AttributeMatching {
+#if os(iOS)
+public typealias AppAttributeMatcher = MobileAppAttributeMatcher
+#elseif os(macOS)
+public typealias AppAttributeMatcher = DesktopAppAttributeMatcher
+#endif
```
Comment (ayoy) [inline Sources/RemoteMessaging/Matchers/AppAttributeMatcher.swift:27]: `installedMacAppStore` is an "app" attribute, so we're adding platform-specific structs for app attribute matching.

File: Sources/RemoteMessaging/Matchers/AppAttributeMatcher.swift [hunk 2]
```diff
@@ -20,7 +20,59 @@ import Foundation
...
+            appVersion: appVersion,
+            isInternalUser: isInternalUser,
+            statisticsStore: statisticsStore,
+            variantManager: variantManager
+        )
+    }
+
+    public func evaluate(matchingAttribute: MatchingAttribute) -> EvaluationResult? {
+        switch matchingAttribute {
+        case let matchingAttribute as IsInstalledMacAppStoreMatchingAttribute:
```
Comment (ayoy) [inline Sources/RemoteMessaging/Matchers/AppAttributeMatcher.swift:63]: Handle `installedMacAppStore` here, and handle other attributes via common matcher.

File: Sources/RemoteMessaging/Matchers/UserAttributeMatcher.swift [hunk 3]
```diff
@@ -95,6 +93,73 @@ public struct MobileUserAttributeMatcher: AttributeMatching {
...
+            isPrivacyProSubscriptionActive: isPrivacyProSubscriptionActive,
+            isPrivacyProSubscriptionExpiring: isPrivacyProSubscriptionExpiring,
+            isPrivacyProSubscriptionExpired: isPrivacyProSubscriptionExpired,
+            dismissedMessageIds: dismissedMessageIds
+        )
+    }
+
+    public func evaluate(matchingAttribute: MatchingAttribute) -> EvaluationResult? {
+        switch matchingAttribute {
+        case let matchingAttribute as PinnedTabsMatchingAttribute:
```
Comment (ayoy) [inline Sources/RemoteMessaging/Matchers/UserAttributeMatcher.swift:146]: Similar to app attributes matcher, handle pinned tabs and custom home page here and use common matcher for the rest.

File: Sources/RemoteMessaging/Model/MatchingAttributes.swift
Comment (ayoy) [inline Sources/RemoteMessaging/Model/MatchingAttributes.swift:1]: Lots of duplicated code was removed from this file via introducing common interfaces for handling matching attributes: `SingleValueMatching`, `NumericRangeMatching` and `StringRangeMatching`.

**Review by jaceklyp — APPROVED** [review]
nice and clean :)

File: Tests/RemoteMessagingTests/Matchers/DesktopUserAttributeMatcherTests.swift [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        mockStatisticsStore.atb = "v105-2"
+        mockStatisticsStore.appRetentionAtb = "v105-44"
+        mockStatisticsStore.searchRetentionAtb = "v105-88"
+        mockStatisticsStore.installDate = dateYesterday
+
+        manager = MockVariantManager(isSupportedReturns: true,
+                                         currentVariant: MockVariant(name: "zo", weight: 44, isIncluded: { return true }, features: [.dummy]))
+        let emailManagerStorage = MockEmailManagerStorage()
+
+        // EmailEnabledMatchingAttribute isSignedIn = true
```
[the remark this answers is not in the document]
Comment (ayoy) [reply Tests/RemoteMessagingTests/Matchers/DesktopUserAttributeMatcherTests.swift:49]: This is an explanation of the code below that sets username and token, which is equivalent to making isSignedIn return true. I'll update this to be more specific.

File: Sources/RemoteMessaging/Matchers/AppAttributeMatcher.swift [hunk 5]
```diff
@@ -20,7 +20,59 @@ import Foundation
...
+
+    public init(
+        bundleId: String,
+        appVersion: String,
+        isInternalUser: Bool,
+        statisticsStore: StatisticsStore,
+        variantManager: VariantManager,
+        isInstalledMacAppStore: Bool
+    ) {
+        self.isInstalledMacAppStore = isInternalUser
```
[the remark this answers is not in the document]
Comment (ayoy) [reply Sources/RemoteMessaging/Matchers/AppAttributeMatcher.swift:50]: Lol it's a bug 😮  should be `= isInstalledMacAppStore`. Thanks for this 🙏

**Pull request merged**